airbyte-cdk/python/airbyte_cdk/sources/file_based/stream/concurrent/cursor/file_based_concurrent_cursor.py

```diff
@@ -0,0 +1,272 @@
+        self._file_to_datetime_history = None
+        self._prev_cursor_value = None
+        self._sync_start = None
+
+        self._file_to_datetime_history = stream_state.get("history", {}) if stream_state else {}
+        self._prev_cursor_value = self._compute_prev_sync_cursor(stream_state)
+        self._sync_start = self._compute_start_time()
+
+    @property
+    def state(self) -> MutableMapping[str, Any]:
+        return self._state
+
+    def observe(self, record: Record) -> None:
+        pass
+
+    def close_partition(self, partition: FileBasedStreamPartition) -> None:
+        with self._pending_files_lock:
+            if self._pending_files is None:
+                raise RuntimeError("Expected pending partitions to be set but it was not. This is unexpected. Please contact Support.")
+
+    def set_pending_partitions(self, partitions: List[FileBasedStreamPartition]) -> None:
+        self._pending_files = {}
+        for partition in partitions:
+            _slice = partition.to_slice()
+            for k in _slice.keys():
+                if k in self._pending_files.keys():
+                    raise RuntimeError(f"Already found file {_slice} in pending files. This is unexpected. Please contact Support.")
+            self._pending_files.update(_slice)
+
+    def _compute_prev_sync_cursor(self, value: Optional[StreamState]) -> Tuple[datetime, str]:
+        if not value:
+            return self.zero_value, ""
+        prev_cursor_str = value.get(self._cursor_field.cursor_field_key) or self._zero_cursor_value()
+        # So if we see a cursor greater than the earliest file, it means that we have likely synced all files.
+        # However, we take the earliest file as the cursor value for the purpose of checking which files to
+        # sync, in case new files have been uploaded in the meantime.
+        # This should be very rare, as it would indicate a race condition where a file with an earlier
+        # last_modified time was uploaded after a file with a later last_modified time. Since last_modified
+        # represents the start time that the file was uploaded, we can usually expect that all previous
+        # files have already been uploaded. If that's the case, they'll be in history and we'll skip
+        # re-uploading them.
+        earliest_file_cursor_value = self._get_cursor_key_from_file(self._compute_earliest_file_in_history())
+        cursor = min(prev_cursor_str, earliest_file_cursor_value)
+        cursor_dt, cursor_uri = cursor.split("_", 1)
+        return datetime.strptime(cursor_dt, self.DATE_TIME_FORMAT), cursor_uri
+
+    def _get_cursor_key_from_file(self, file: Optional[RemoteFile]) -> str:
+        if file:
+            return f"{datetime.strftime(file.last_modified, self.DATE_TIME_FORMAT)}_{file.uri}"
+        return self._zero_cursor_value()
+
+    def _zero_cursor_value(self) -> str:
+        return f"{self.zero_value.strftime(self.DATE_TIME_FORMAT)}_"
```

Comment:
Why the `_` at the end of the string?

Reply:
This might seem dumb in terms of code because it doesn't do anything but should we have a `return f"{self.zero_value.strftime(self.DATE_TIME_FORMAT)}_{self._NULL_FILE}"` where `_NULL_FILE == ""` or something like that to make sure we understand why the `_`

Reply:
👍 makes sense to me! Updated.

Reply:
👍 makes sense to me!